File: go/internal/sol/p0093/p0093.go
package p0093

/*
combination of numberes:
  nCk = C(n,k) = C(9,4) = 9*8*7*6 / 4*3*2*1 = 126

arithmetic operations (fourOps):
  commutative:
    addition: X + Y
    multiplication: X * Y
  no-commutative:
    subtraction:  X - Y, Y - X
    division: X / Y, Y / X

patterns:
  A, B, C, D: numbers
    4! = 24

  [1] ((A op B) op C) op D
       ^^^^^^^^
  [2] (A op B) op (C op D)
      ^^^^^^^^    ^^^^^^^^
	  ^^^^^^^^^^^^^^^^^^^^

  ^^^: The order of the two terms is irrelevant
       because fourOps() also considers the reverse order.
*/

import (
	"math/big"
	"slices"
	"strconv"

	"pe-solver/internal/mylib"
)

var ratZero = big.NewRat(0, 1)

func fourOps(x1, x2 *big.Rat) []*big.Rat {
	result := make([]*big.Rat, 0, 6)

	if x1.Cmp(ratZero) == 0 {
		result = append(result, big.NewRat(0, 1))
		result = append(result, x2)
		result = append(result, new(big.Rat).Neg(x2))
	} else if x2.Cmp(ratZero) == 0 {
		result = append(result, big.NewRat(0, 1))
		result = append(result, x1)
		result = append(result, new(big.Rat).Neg(x1))
	} else {
		result = append(result, new(big.Rat).Add(x1, x2))
		result = append(result, new(big.Rat).Mul(x1, x2))
		result = append(result, new(big.Rat).Sub(x1, x2))
		result = append(result, new(big.Rat).Sub(x2, x1))
		result = append(result, new(big.Rat).Mul(x2, new(big.Rat).Inv(x1)))
		result = append(result, new(big.Rat).Mul(x1, new(big.Rat).Inv(x2)))
	}

	return result
}

func case1(d1, d2 *big.Rat, rest []*big.Rat) []*big.Rat {
	result := make([]*big.Rat, 0)
	for ab := range slices.Values(fourOps(d1, d2)) {
		// c: rest[0], d: rest[1]
		for abc := range slices.Values(fourOps(ab, rest[0])) {
			result = append(result, fourOps(abc, rest[1])...)
		}
		// c: rest[1], d: rest[0]
		for abc := range slices.Values(fourOps(ab, rest[1])) {
			result = append(result, fourOps(abc, rest[0])...)
		}
	}

	return result
}

func case2(d1, d2 *big.Rat, rest []*big.Rat) []*big.Rat {
	result := make([]*big.Rat, 0)
	ab := fourOps(d1, d2)
	cd := fourOps(rest[0], rest[1])

	for pair := range mylib.CartesianProduct(ab, cd) {
		result = append(result, fourOps(pair[0], pair[1])...)
	}

	return result
}

func makeNumbers(lst []*big.Rat) map[int]struct{} {
	nSet := map[int]struct{}{}

	for i := 0; i < len(lst); i++ {
		for j := i + 1; j < len(lst); j++ {
			rest := make([]*big.Rat, 0, 2)
			for k := 0; k < len(lst); k++ {
				if k != i && k != j {
					rest = append(rest, lst[k])
				}
			}

			for x := range slices.Values(case1(lst[i], lst[j], rest)) {
				if x.IsInt() {
					nSet[int(x.Num().Int64())] = struct{}{}
				}
			}
			for x := range slices.Values(case2(lst[i], lst[j], rest)) {
				if x.IsInt() {
					nSet[int(x.Num().Int64())] = struct{}{}
				}
			}
		}
	}

	return nSet
}

func countConsecNumbers(lst []int) int {
	ratLst := make([]*big.Rat, 0)
	for x := range slices.Values(lst) {
		ratLst = append(ratLst, big.NewRat(int64(x), 1))
	}

	nSet := makeNumbers(ratLst)
	for cnt := 1; ; cnt++ {
		if _, ok := nSet[cnt]; ok {
			continue
		}
		return cnt - 1
	}
}

func compute() string {
	var maxCount int
	var nums []int

	for lst := range mylib.Combinations([]int{1, 2, 3, 4, 5, 6, 7, 8, 9}, 4) {
		tmp := countConsecNumbers(lst)
		if tmp > maxCount {
			maxCount = tmp
			nums = lst
		}
	}

	var result int
	for x := range slices.Values(nums) {
		result = result*10 + x
	}

	return strconv.FormatInt(int64(result), 10)
}

func Solve() string {
	return compute()
}

File: haskell/src/Sol/P0093.hs
module Sol.P0093 (compute, solve) where

{-
This implementation is little slow. The following is a result on my Raspberry Pi 4.

% cabal v2-run pe-solver -- 93
[Problem 93]
Answer: 1258
Elapsed time: 0.843592 sec.
-}

{-
arithmetic operations (fourArithmeticOps):
  commutative:
    addition: X + Y
    multiplication: X * Y
  no-commutative:
    subtraction:  X - Y, Y - X
    division: X / Y, Y / X

patterns:
  d1, d2, d3, d4: numbers

  [case 1]
     ((d1 OP d2) OP d3) OP d4
      ^^^----^^^
     ^^^^^^^^^^^----^^^
     ^^^^^^^^^^^^^^^^^^----^^

     ((d1 OP d2) OP d4) OP d3
      ^^^----^^^
     ^^^^^^^^^^^----^^^
     ^^^^^^^^^^^^^^^^^^----^^
  [case 2]
     (d1 OP d2) OP (d3 OP d4)
     ^^^----^^^    ^^^----^^^
     ^^^^^^^^^^----^^^^^^^^^^

  ^^-^^: We can ignore the order of the two terms because
         fourArithmeticOps() considers no-commutative operations.
-}

import Control.Arrow ((&&&))
import Data.Function (on)
import Data.List (maximumBy, (\\))
import Data.Ratio (Ratio, denominator, numerator)

import qualified Data.IntSet as S (fromList, member)

import Mylib.Combinatorics (combinations)
import Mylib.Util (headExn, tailExn)

fourArithmeticOps :: Ratio Int -> Ratio Int -> [Ratio Int]
fourArithmeticOps x y
    | x == 0 =
        [negate y, 0, y]
    | y == 0 =
        [negate x, 0, x]
    | otherwise =
        [x + y, x - y, y - x, x * y, x / y, y / x]

{- HLINT ignore case_1 "Use head" -}
case_1 :: ([Ratio Int], [Ratio Int]) -> [Ratio Int]
case_1 (lst1, lst2) =
    let res1 = d1d2 >>= fourArithmeticOps d3 >>= fourArithmeticOps d4
        res2 = d1d2 >>= fourArithmeticOps d4 >>= fourArithmeticOps d3
     in res1 ++ res2
  where
    d1d2 = fourArithmeticOps (lst1 !! 0) (lst1 !! 1)
    d3 = lst2 !! 0
    d4 = lst2 !! 1

{- HLINT ignore case_2 "Use head" -}
case_2 :: ([Ratio Int], [Ratio Int]) -> [Ratio Int]
case_2 (lst1, lst2) =
    concat $ liftA2 fourArithmeticOps d1d2 d3d4
  where
    d1d2 = fourArithmeticOps (lst1 !! 0) (lst1 !! 1)
    d3d4 = fourArithmeticOps (lst2 !! 0) (lst2 !! 1)

makeIntNumbers :: [Ratio Int] -> [Int]
makeIntNumbers =
    map numerator . filter ((== 1) . denominator) . aux
  where
    aux :: [Ratio Int] -> [Ratio Int]
    aux lst =
        l1 ++ l2
      where
        l1 = case_1 . (id &&& (lst \\)) =<< combinations 2 lst
        l2 = case_2 . (id &&& (lst \\)) =<< map (: [headExn lst]) (tailExn lst)

countConsecNumbers :: [Ratio Int] -> Int
countConsecNumbers lst =
    go 1
  where
    set = S.fromList (makeIntNumbers lst)

    go :: Int -> Int
    go cnt
        | S.member cnt set =
            go (succ cnt)
        | otherwise =
            pred cnt

compute :: String
compute =
    concatMap (show . numerator)
        . snd
        . maximumBy (compare `on` fst)
        . map (countConsecNumbers &&& id)
        $ combinations 4 [1 .. 9 :: Ratio Int]

solve :: String
solve = compute

{-
The following implementation is about twice slower than the above. I wonder why.

compute :: String
compute =
    concatMap (show . numerator)
        . maximumBy (compare `on` countConsecNumbers)
        $ combinations 4 [1 .. 9 :: Ratio Int]

% cabal v2-run pe-solver -- 93
[Problem 93]
Answer: 1258
Elapsed time: 1.699603 sec.
-}
